statuslistener/statuslistener_test.go

```diff
@@ -95,13 +95,21 @@ func PopulateDateFieldsFrom(resource interface{}) DateFields {
 		dateFields.UpdatedAt = typedResource.UpdatedAt
 	case *m.Application:
 		dateFields.CreatedAt = typedResource.CreatedAt
-		dateFields.LastAvailableAt = *typedResource.LastAvailableAt
-		dateFields.LastCheckedAt = *typedResource.LastCheckedAt
+		if typedResource.LastAvailableAt != nil {
```

Comment:
Do we need the same change for case `*m.Source`?

Reply:
No, because when the status listener receives an availability status message, it updates the `LastAvailableAt` —if the resource's status is `available`— and `LastCheckedAt` in the `attributesForUpdate` function:

https://github.com/RedHatInsights/sources-api-go/blob/e70a41501ac415d28047f0dca80b59130fc05be4/statuslistener/statuslistener.go#L144-L160

No, and let me explain: in the case of these tests, we load the JSON in memory and we update the timestamps to the ones that the resources have in the database, because otherwise when the status listener generates the bulk message, it would have the latest timestamps from the database, but when comparing those timestamps to the ones hard coded in the JSON files, the comparison would fail. So, in order to avoid that, we fetch the resource and update the timestamps of the JSON:

https://github.com/RedHatInsights/sources-api-go/blob/e70a41501ac415d28047f0dca80b59130fc05be4/statuslistener/statuslistener_test.go#L308-L316

The `LastAvailableAt` and `LastCheckedAt` attributes are updated by the status listener upon reception of an availability status message:

https://github.com/RedHatInsights/sources-api-go/blob/e70a41501ac415d28047f0dca80b59130fc05be4/statuslistener/statuslistener.go#L144-L160

So, why do we need to set the `if`s only for the `Application` and `Endpoint` resources? Because they are the only ones that can be a subresource of a `Source`. The problem comes when we are trying to test the bulk message for the `Source` resource, because we fetch the source and all its subresources. However, those subresources —`Application`s and `Endpoint`s— have null `LastAvailableAt` and `LastCheckedAt` attributes because the status listener only received a status message for the `Source`, not for its subresources.

That explains the pointer dereference error.

Reply:
Thanks for explanation !!!